### ISSUE TITLE
the operation of attributes not defined

### ISSUE THREAD

**Issue by ddzrc** [open-telemetry/opentelemetry-specification#1857]
**What are you trying to achieve?**
I tried to implement open-telemetry in the perl language, but I found that the definition of attributes was not comprehensive enough:
1. Attributes only defines attributes, not Include methods.
2. The implementation of other languages is also inconsistent.
3. Span can set attributes, indicating that users can operate attributes

What did you expect to see?
I think it is very necessary to define the atrribute method

**Comment by Oberon00** [comment]
The methods to define attributes are defined, e.g. for spans here:
* Set at creation: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#span-creation
* Set with `SetAttribute`: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#set-attributes

The attribute type itself is defined at https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/common.md#attributes

What other attribute methods do you expect? Are you maybe talking about how the attributes can be read back? That is indeed mostly language-specific but the basics are defined here: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#additional-span-interfaces (note that the these interfaces should only be available to implementations of Exporter.Export, SpanProcessor.OnStart and SpanProcessor.OnEnd).

**Comment by ddzrc** [comment]
before call set-attributes， we should create Attributes and add key - value into Attributes， in sdk should read key-value from Attributes

**Comment by ddzrc** [comment]
> The methods to define attributes are defined, e.g. for spans here:
> 
> * Set at creation: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#span-creation
> * Set with `SetAttribute`: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#set-attributes
> 
> The attribute type itself is defined at https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/common.md#attributes
> 
> What other attribute methods do you expect? Are you maybe talking about how the attributes can be read back? That is indeed mostly language-specific but the basics are defined here: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#additional-span-interfaces (note that the these interfaces should only be available to implementations of Exporter.Export, SpanProcessor.OnStart and SpanProcessor.OnEnd).
before call set-attributes， we should create Attributes and add key - value into Attributes， in sdk should read key-value from Attributes

**Comment by Oberon00** [comment]
> we should create Attributes 

That is up to the language. Usually, `Attribute` is not a separate type. E.g. `SetAttribute` may just be a set of methods like `SetAttribute(string key, int value)`, `SetAttribute(string key, string value)`, etc.

**Comment by Oberon00** [comment]
Thinking about it, it would actually be quite helpful to add a note to the Attribute spec that contrary to most other concepts (Span, Tracer, SpanContext, ...), Attributes should be seen as a concept that is not represented by a dedicated object/type but usually implemented by having separate key, value arguments or a key -> value dictionary/map.

**Comment by ddzrc** [comment]
![image](https://user-images.githubusercontent.com/39239990/128672464-15a3f041-80ab-45dc-9dd7-95833697e2a9.png)
the Attributes param of  span create ， can been dictionary/map？

**Comment by ddzrc** [comment]
> Thinking about it, it would actually be quite helpful to add a note to the Attribute spec that contrary to most other concepts (Span, Tracer, SpanContext, ...), Attributes should be seen as a concept that is not represented by a dedicated object/type but usually implemented by having separate key, value arguments or a key -> value dictionary/map.

----------------------------------------------------------------------------------------------------------------------

![image](https://user-images.githubusercontent.com/39239990/128672464-15a3f041-80ab-45dc-9dd7-95833697e2a9.png)

**the Attributes param of  span create ， can been dictionary/map？**

**Comment by Oberon00** [comment]
Yes, it can be a dictionary/map or whatever makes sense in your language. E.g. in Java the Builder pattern is used, so for starting a span you call `Tracer.spanBuilder(name)` which returns a `SpanBuilder` object that has a `SetAttribute(key, value)` method (among others) as well as a `start()` method that returns the created Span.

**Comment by ddzrc** [comment]
> Yes, it can be a dictionary/map or whatever makes sense in your language. E.g. in Java the Builder pattern is used, so for starting a span you call `Tracer.spanBuilder(name)` which returns a `SpanBuilder` object that has a `SetAttribute(key, value)` method (among others) as well as a `start()` method that returns the created Span.

Thanks

**Comment by ddzrc** [comment]
> Yes, it can be a dictionary/map or whatever makes sense in your language. E.g. in Java the Builder pattern is used, so for starting a span you call `Tracer.spanBuilder(name)` which returns a `SpanBuilder` object that has a `SetAttribute(key, value)` method (among others) as well as a `start()` method that returns the created Span.

------------------------------------------------------------------------------------------------------------------------------------------
As part of api, does it need to be more specific?
